orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -419,4 +388,12 @@ def _get_forward_graph(module, *inputs, **kwargs):
                           training=torch.onnx.TrainingMode.TRAINING,
                           dynamic_axes=dynamic_axes)
 
-        return onnx.load_model_from_string(f.getvalue())
+        model = onnx.load_model_from_string(f.getvalue())
+
+        # Models without parameters are not allowed
+        pt_count_params = sum([1 for param in module.parameters()])
+        onnx_count_params = sum([1 for param in model.graph.initializer])
+        if pt_count_params < 1 or onnx_count_params < 1:
+            raise RuntimeError('ORTModule only supports model with at least one trainable parameter')
```

Comment:
I believe the snippet above is more of a workaround than how we want to support ORTModule. Maybe @mrry can comment
If we really want to use such architecture, yes, we would have to remove the wrap around ReLU

Reply:
We should aim to support as many module types as we can in ORT. Since we're aiming for a seamless experience with PyTorch, the behavior should match whatever PyTorch does in this case.

For `nn.ReLU()`, I would guess it runs on whatever device the input tensor is placed on, and the `.to(device)` call on the parameterless module is a no-op. So this is kind of a strange example, but it's still worth supporting.

Reply:
Can https://github.com/microsoft/onnxruntime/pull/6417 be reviewed/merged first? I will try to reuse some of its refactoring to support this

One issue that I see is that `InferenceSession` is created when the onnx graph is exported, before the training loop is actually executed.

One way of supporting this is using cpu as default when there is no device in the model and invalidate the inferencesession when a self._device is none and user data has device on it.

Reply:
@mrry I've made some changes to allow models like `model = torch.nn.ReLU()` to  tried passing torch.nn.ReLU() to ORT backend. The idea is that if the model doesnt have a device in the initialization, we catch the device from input in the first forward, when we export the model to onnx.

The issue arises when we try to create a InferenceSession for a model with ReLU only:

```
/home/thiagofc/miniconda3/envs/ort_py37/lib/python3.7/site-packages/torch/nn/modules/module.py:744: in _call_impl
    result = self.forward(*input, **kwargs)
onnxruntime/training/ortmodule.py:207: in forward
    self._onnx_backward.SerializeToString(), providers=providers, provider_options=provider_options)
onnxruntime/capi/onnxruntime_inference_collection.py:206: in __init__
    self._create_inference_session(providers, provider_options)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <onnxruntime.capi.onnxruntime_inference_collection.InferenceSession object at 0x7efb803fc890>, providers = ['CPUExecutionProvider'], provider_options = [{}]

    def _create_inference_session(self, providers, provider_options):
        # Tensorrt can fall back to CUDA. All others fall back to CPU.
        if 'TensorrtExecutionProvider' in C.get_available_providers():
            self._fallback_providers = ['CUDAExecutionProvider', 'CPUExecutionProvider']
        else:
            self._fallback_providers = ['CPUExecutionProvider']
    
        session_options = self._sess_options if self._sess_options else C.get_default_session_options()
        if self._model_path:
            sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
        else:
            sess = C.InferenceSession(session_options, self._model_bytes, False, self._read_config_from_model)
    
        # initialize the C++ InferenceSession
>       sess.initialize_session(providers or [], provider_options or [])
E       onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Exception during initialization: /home/thiagofc/dev/github/onnxruntime/onnxruntime/core/framework/session_state_utils.cc:101 onnxruntime::common::Status onnxruntime::session_state_utils::SaveInitializedTensors(const onnxruntime::Env&, const std::__cxx11::basic_string<char>&, const onnxruntime::GraphViewer&, const OrtMemoryInfo&, const onnxruntime::OrtValueNameIdxMap&, const std::vector<int>&, onnxruntime::ITensorAllocator&, const std::function<onnxruntime::common::Status(int, const OrtValue&, const onnxruntime::OrtCallback&, bool)>&, const onnxruntime::logging::Logger&, const onnxruntime::DataTransferManager&, const onnxruntime::ExecutionPlanBase&, const onnxruntime::SessionOptions&) ort_value_name_idx_map.MaxIdx() > -1 was false. OrtValue indexes should have been populated.
```
I don't know the details yet, but it might be there is no "initializers" at `onnxruntime::session_state_utils::SaveInitializedTensors(...,ort_value_name_idx_map,...)`, which is a problem for the backend

Reply:
I've discussed with @SherlockNoMad and he mentioned the error above is a corner case and there might be other places that assume initializer do exist. My idea was to use PyTorch `forward` whenever we detect model without trainable parameters. What do you think @mrry ?

Reply:
This seems like an unjustified assumption for the code to have, and I'd like to probe it further :). Can you try patching https://github.com/microsoft/onnxruntime/commit/5e11a8254c973541093c0bd7bc0c32917ac2ce42 and see if it unblocks things?

Reply:
For this PR, 4 models were proposed to cover all corner cases

```
@pytest.mark.parametrize("device", ['cuda', 'cpu'])
def test_model_without_parameters(device):
    class Net(torch.nn.Module):
        def forward(self, x):
            return x

    model = Net().to(device)
    model = ORTModule(model).to(device)
    _ = model(torch.tensor(1.))

@pytest.mark.parametrize("device", ['cuda', 'cpu'])
def test_model_without_trainable_parameters(device):

    model = torch.nn.ReLU()
    model = ORTModule(model).to(device)

    batch_size = 5
    nb_classes = 2
    in_features = 10
    input = torch.randn(batch_size, in_features, device=device)
    target = torch.empty(batch_size, dtype=torch.long, device=device).random_(nb_classes)
    _ = model(input)

@pytest.mark.parametrize("device", ['cuda', 'cpu'])
def test_custom_model_without_trainable_parameters(device):
    class Net(torch.nn.Module):
        def forward(self, x):
            return torch.nn.ReLU(x)

    model = Net()
    model = ORTModule(model).to(device)
    model.to(device)
    _ = model(torch.tensor(1.).to(device))

@pytest.mark.parametrize("device", ['cuda', 'cpu'])
def test_model_with_unused_trainable_parameters(device):
    class Net(torch.nn.Module):
        def __init__(self, input_size, hidden_size, num_classes):
            super(Net, self).__init__()
            self.fc1 = torch.nn.Linear(input_size, hidden_size)
            self.relu = torch.nn.ReLU()
            self.fc2 = torch.nn.Linear(hidden_size, num_classes)

        def forward(self, input1):
            return input1

    model = Net(784, 500, 10)
    model = ORTModule(model).to(device)
    model(torch.tensor(1.).to(device))
```
* After your patch, `test_model_without_trainable_parameters` succeeds. Thanks!

* `test_custom_model_without_trainable_parameters` is non-exportable because `Only tuples, lists and Variables supported as JIT inputs/outputs. Dictionaries and strings are also accepted but their usage is not recommended. But got unsupported type ReLU`. It is ok to crash in this scenario, but there are alternatives we can discuss at the end of this comment [1]

* `test_model_without_parameters` fails because the gradient graph builder can't find the producer of the output X, which is the very same input X. This The reason is that ONNX exporter doesn't add an Identity node to connect input x to output x, but instead connect the two directly and ORT doesn't like that a bit

![image](https://user-images.githubusercontent.com/5469809/106035224-b6957900-6088-11eb-9f8a-894de093d730.png)

```
    def forward(self, *inputs, **kwargs):
        '''Forward pass starts here and continues at `_ORTModuleFunction.forward`
    
        ONNX model is exported the first time this method is executed.
        Next, a full training graph is splitted in forward and backward graph which are used
        to instantiate ONNX Runtime InferenceSession`s
        '''
        if not self._onnx_forward or self._require_export:
            self._require_export = False
            if not self._device:
                self._device = _utils.get_device_index_from_input_args_kwargs(self._original_module, *inputs, **kwargs)
                if not self._device:
                    raise RuntimeError('A device must be specified in the model or data!')
    
            self._onnx_training, self._use_pytorch_forward = ORTModule._get_forward_graph(self._original_module, *inputs, **kwargs)
            if not self._use_pytorch_forward:
                if self._save_onnx:
                    onnx.save(self._onnx_training, self._save_onnx_prefix + '_full_training.onnx')
    
                # TODO: PyTorch exporter bug: changes the initializer order
                initializer_names = [p[0] for p in self._original_module.named_parameters()]
    
                # Build full training graph and split in forward/backward
                grad_builder_config = C.ModuleGradientGraphBuilderConfiguration()
                grad_builder_config.initializer_names_to_train = initializer_names
                grad_builder_config.input_names_require_grad = []
                self._module_gradient_graph_builder = C.ModuleGradientGraphBuilder()
                self._module_gradient_graph_builder.initialize(self._onnx_training.SerializeToString(), grad_builder_config)
    
        # Perform shape inference and re-split forward/backward graph for bacthes with different shapes
        if not self._use_pytorch_forward:
            new_input_shape = [list(input.size()) for input in inputs if input is not None]
            if self._current_input_shape is None or self._current_input_shape != new_input_shape:
                self._current_input_shape = new_input_shape
>               self._module_gradient_graph_builder.build_and_split(self._current_input_shape)
E               RuntimeError: /github/onnxruntime/orttraining/orttraining/core/framework/gradient_graph_builder.cc:58 onnxruntime::training::GradientGraphBuilder::GradientGraphBuilder(onnxruntime::Graph*, const std::unordered_set<std::__cxx11::basic_string<char> >&, const std::unordered_set<std::__cxx11::basic_string<char> >&, const string&, const onnxruntime::training::GradientGraphConfiguration&, const onnxruntime::logging::Logger&) x couldn't find the producer node.
```
I fixed that by issuing a warning when the producer node is not found, but allowed gradient build to move on.

Finally, `test_model_with_unused_trainable_parameters` failed during graph splitting because the unused weights are not found in the ONNX graph (because unused nodes are not exported), but they are listed at  `C.ModuleGradientGraphBuilderConfiguration.initializer_names_to_train` as valid initializers. I worked this around by removing initializers that are not listed on the onnx initializer list too. After this fix, the same issue as before showed up. ORT core didnt like the fact that the input was directly passed to the output without a node in between

[1] The alternative I mentioned before was to issue an informative warning message describing what failed (including stack) and fallback to PyTorch forward() to allow user to train. That could be a good short term fix for scenarios that we do not support but are used in real life

What do you think?

Reply:
Thanks for digging into this, Thiago! I'm not sure I understand why `test_custom_model_without_trainable_parameters` is non-exportable... from the error message it sounds like the module is being passed where the input tensor is expected (although perhaps I'm misunderstanding). The following worked for me, which is what I *think* is happening under the covers:

```
>>> import torch
>>> model = torch.nn.ReLU()
>>> torch.onnx.export(model, torch.Tensor([1.0, 2.0, 3.0]), "relu.onnx")
```

Re: falling back to PyTorch, I'd prefer not to add code for doing this just yet. Instead, let's track these cases as bugs, and we'll revisit when we're closer to release.

Reply:
There is no fallback to PyTorch in the latest version

- `test_custom_model_without_trainable_parameters` cannot be exported because the `forward` output is a `Relu` object, not actual data
- `test_model_without_trainable_parameters` and `test_model_without_parameters` and `test_model_with_unused_trainable_parameters` are handled by ORT core
- `test_model_with_unused_trainable_parameters` was also handled on the frontend

Reply:
Backend was updated to allow models without trainable weights